dashboard/app/assets/javascripts/header.js.erb

```diff
@@ -2,110 +2,575 @@
+      $el.off('click', this.onClick);
+      $el.on('destroyed', this.onDestroy);
+
+      this.componentDidUnmount();
+    }
+
+    // Return a default state for this component
+    UIComponent.prototype.getInitialState = function() { }
+
+    // Extend/merge my state and possibly re-render
+    UIComponent.prototype.setState = function(state) {
+      if (!this.isUpdating) {
+        // Allow things to settle - no need to call more than once per execution
+        this.isUpdating = setTimeout( this.onUpdate, 1 );
+      }
+
+      // Merge this into our state
+      for (var key in state) {
+        if (__hasProp.call(state, key))
+          this.state[key] = state[key];
+      }
+    }
+
+    // Replace my state and possibly re-render
+    UIComponent.prototype.replaceState = function(state) {
+      if (!this.isUpdating) {
+        // Allow things to settle - no need to call more than once per execution
+        this.isUpdating = setTimeout( this.onUpdate, 1 );
+      }
+
+      this.state = state || {};
+    }
+
+    // My state has been changed, and we've let all other processes yield, so we know
+    // there are no more changes coming.  Time to update the DOM.
+    UIComponent.prototype.onUpdate = function(ev) {
+      // Don't do anything if no longer mounted
+      if (!this.isMounted())
+        return;
+
+      this.isMounting ? this.componentWillMount() : this.componentWillUpdate();
+
+      this.render();
+      delete this.isUpdating;
+
+      this.isMounting ? this.componentDidMount() : this.componentDidUpdate();
+      delete this.isMounting;
+    }
+
+    // Render into this element, given some state
+    // There is no default rendering
+    UIComponent.prototype.render = function() { }
+
+    // Handle a click
+    // There is no default click behavior
+    UIComponent.prototype.onClick = function(ev) { }
+
+    // Mount/unmount notifications
+    UIComponent.prototype.componentWillMount = function() { }
+    UIComponent.prototype.componentDidMount = function() { }
+    UIComponent.prototype.componentWillUnmount = function() { }
+    UIComponent.prototype.componentDidUnmount = function() { }
+    UIComponent.prototype.componentWillUpdate = function() { }
+    UIComponent.prototype.componentDidUpdate = function() { }
+    UIComponent.prototype.isMounted = function() {
+      return this.$el !== null;
+    }
+
+    return UIComponent;
+  })();
+
+
+
+
+
+  // class UIRouter: this is an incredibly simple router with our routes baked into it.
+  // This will be completely replaced when we choose a router.
+  UIRouter = (function() {
+
+    function UIRouter() { }
+
+    UIRouter.prototype.route = function() {
+      var re, parts,
+        url = window.location.href;
+
+      // s/:script_id/level/:level_id
+      // (only applies to the 20-hour course)
+      re = /\/s\/\d+\/level\/(\d+)$/i;
+      parts = url.match(re);
+      if (parts) {
+        window.headerProgress = new HeaderProgress( $('.header-wrapper'), {
```

Comment:
Feels funny to have an element reference here - in a real router this wouldn't happen.